### DISCUSSION TITLE
Support of decimal timing points

### DISCUSSION THREAD

**Discussion by TengZhang19** [ppy/osu#14834] [Ideas]
After the release of "Stable 20210920", some players and mappers already reported the issue with some sv/gimmick mania maps in which decimal timing points are used for better visual effect or a certain kind of cool barline gimmick.

examples of maps:
https://osu.ppy.sh/beatmapsets/1237210#mania/2571594 
https://osu.ppy.sh/beatmapsets/1226110#mania/2549753
and many others

what they look like before the update:
https://www.youtube.com/watch?v=V5OMJpN00Js
https://www.youtube.com/watch?v=-m9om0D-WgQ&t=73s

Personally, the old osu timing system is a quite great one compared to other rhythm games, it could handle very complex and accurate timings with fast processing time, which is also why I enjoy making cool sv/gimmick maps by osu. The system has great potential since

1. it could be used to create sv with better quality. The limitation of inherited points(0.1~10) make it impossible to create statics effect (sv=0) or to avoid broken frames (sv=10 is still too slow for teleports)
2. the cool creation of barline gimmicks (just like those in the maps listed above) 

The support of decimal timing points is quite an interesting feature, looks like it conflicts with a new feature in Lazer so it's simplified, but I think many players and mappers want it back. 

I think it would be cool if this feature could be compatible with the new feature of Lazer in some way, If it's possible to make it in Lazer then the issue with stable maps won't be a problem anymore, and those old cool maps won't be broken and could be successfully transferred to osu!Lazer

**Comment by mcendu** [comment]
What do you mean by "decimal"?

**Reply by frenzibyte** [reply]
Floating-point numbers, in other words.

**Comment by bdach** [comment]
For context, reasoning for doing this can be found here: https://github.com/ppy/osu/pull/14619#issuecomment-918906748

**Comment by datoujia** [comment]
This is not only related to the measure line gimmick, but some sv whick have decimal offset and large speed value in mania mode. For example, https://osu.ppy.sh/b/2716101
Look 01:14:356 - , I used 9.1x speed for 3.6 ms, and used 0.1x speed for 32.4ms. Ideally, average speed is (9.1*3.6+0.1*32.4)/36=1. It is called sightreadable sv, which ensure that people can play it easily without memorization. In latest osu (20210920), the length of 9.1x is round to 3ms, and the length of 0.1x is round to 33ms. It change the average speed to (9.1*3+0.1*33)/36=0.85. 15% average speed change is very very prominent and not negligible.

**Reply by datoujia** [reply]
Sorry Idk why asterisk are hidden
(9.13.6+0.132.4)/36=1 →  (9.1×3.6+0.1×32.4)/36=1
(9.13+0.133)/36=0.85 →  (9.1×3+0.1×33)/36=0.85

**Reply by datoujia** [reply]
Some sv which have interval time < 1ms are broken and unplayable. Just like 04:04:767 - in this map, and 00:49:940 - in the loved mapset 1237210.

**Reply by mcendu** [reply]
Asterisk is a formatting mark here that denotes italics and bolds. Escaping can be done with backslash.

Example:
`*italic*`: *italic*
`**bold**`: **bold**
`\*asterisks\*`: \*asterisks\*

**Comment by Opposite34** [comment]
Seems like this was reverted back on the cutting edge version on Sept 24th: https://osu.ppy.sh/home/changelog/cuttingedge/20210924
So, I think the issue on stable should be fixed soon. 

However, I think this still needs an implementation somehow to work on Lazer as well as it seems like the change from using double to int was done originally to support the Lazer editor.

**Reply by peppy** [reply]
This whole discussion is misguided. Have you actually tested beatmaps on lazer? If you have and they don't work as expected, please open a new discussion. This code base already supports doubles for all time values, so if there's an issue with handling it is likely elsewhere. And may or may not be supported going forward depending on how convoluted the handling is.
